Cubical/Categories/Site/Sheafification/Base.agda

```diff
@@ -0,0 +1,126 @@
+{-# OPTIONS --safe #-}
+module Cubical.Categories.Site.Sheafification.Base where
+
+-- Construction of the sheafification of a presheaf on a site
+-- using a quotient inductive type (QIT).
+
+-- This is inspired by the construction of the sheafification (for finite coverings) in:
+-- * E. Palmgren, S.J. Vickers, "Partial Horn logic and cartesian categories"
+
+open import Cubical.Foundations.Prelude
+open import Cubical.Foundations.Structure
+open import Cubical.Foundations.Function using (_∘_)
+
+open import Cubical.HITs.PropositionalTruncation using (∣_∣₁)
+
+open import Cubical.Data.Sigma
+
+open import Cubical.Functions.Surjection
+open import Cubical.Functions.Embedding
+
+open import Cubical.Categories.Category
+open import Cubical.Categories.Presheaf
+open import Cubical.Categories.Functor
+
+open import Cubical.Categories.Site.Cover
+open import Cubical.Categories.Site.Coverage
+open import Cubical.Categories.Site.Sheaf
+
+
+module Sheafification
+  {ℓ ℓ' ℓcov ℓpat : Level}
+  {C : Category ℓ ℓ'}
+  (J : Coverage C ℓcov ℓpat)
+  {ℓP : Level}
+  (P : Presheaf C ℓP)
+  where
+
+  open Category C hiding (_∘_)
+  open Coverage J
+
+  data ⟨F⟅_⟆⟩ : ob → Type (ℓ-max ℓ (ℓ-max ℓ' (ℓ-max ℓcov (ℓ-max ℓpat ℓP)))) where
+
+    trunc : {c : ob} → isSet ⟨F⟅ c ⟆⟩
+
+    restrict : {c d : ob} → (f : Hom[ d , c ]) → ⟨F⟅ c ⟆⟩ → ⟨F⟅ d ⟆⟩
+
+    restrictId :
+      {c : ob} →
+      (x : ⟨F⟅ c ⟆⟩) →
+      restrict id x ≡ x
+    restrictRestrict :
+      {c d e : ob} →
+      (f : Hom[ d , c ]) →
+      (g : Hom[ e , d ]) →
+      (x : ⟨F⟅ c ⟆⟩) →
+      restrict (g ⋆ f) x ≡ restrict g (restrict f x)
+
+    η⟦⟧ : {c : ob} → (x : ⟨ P ⟅ c ⟆ ⟩) → ⟨F⟅ c ⟆⟩
+
+    ηNatural :
+      {c d : ob} →
+      (f : Hom[ d , c ]) →
+      (x : ⟨ P ⟅ c ⟆ ⟩) →
+      η⟦⟧ ((P ⟪ f ⟫) x) ≡ restrict f (η⟦⟧ x)
+
+    sep :
+      {c : ob} →
+      (cover : ⟨ covers c ⟩) →
+      let cov = str (covers c) cover in
+      (x y : ⟨F⟅ c ⟆⟩) →
+      (i~j :
+        (patch : ⟨ cov ⟩) →
+        restrict (patchArr cov patch) x ≡ restrict (patchArr cov patch) y) →
+      x ≡ y
+
+    amalgamate :
+      let
+      -- Is there any way to make this definition now and reuse it later?
+      F : Presheaf C _
+      F = record
+        { F-ob = λ c → ⟨F⟅ c ⟆⟩ , trunc
+        ; F-hom = restrict
+        ; F-id = funExt restrictId
+        ; F-seq = λ f g → funExt (restrictRestrict f g)
+        }
+      in
+      {c : ob} →
+      (cover : ⟨ covers c ⟩) →
+      let cov = str (covers c) cover in
```

Comment:
Perhaps it would be nice to have some syntax for this construction. Something like `∀[x in Fam] P x ≔ ∀ (i : ⟨ Fam ⟩) → P (str Fam i)` where `Fam : TypeWithStr ℓ λ I → I → A` and `P : A → Type`

Reply:
I think this is out of scope for this PR.

In principle I would like to have a file dedicated to "families" of things (with a bundled index type). But I wouldn't know where to put it in the library. Or how to find places in the library where it could already be used. Or how to handle potential variants like "a family indexed by an h-set" nicely.